addons/binding/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/motionsensor/DLinkMotionSensorCommunication.java

```diff
@@ -0,0 +1,288 @@
+
+    private boolean loginSuccess;
+    private boolean detectSuccess;
+
+    private long prevDetection;
+    private long lastDetection;
+
+    private final ScheduledFuture<?> detectFuture;
+
+    private boolean online = true;
+    private DeviceStatus status = DeviceStatus.INITIALISING;
+
+    /**
+     * Inform the listener if motion is detected
+     */
+    private final Runnable detect = new Runnable() {
+        @Override
+        public void run() {
+            boolean updateStatus = false;
+
+            switch (status) {
+                case INITIALISING:
+                    online = false;
+                    updateStatus = true;
+                    // FALL-THROUGH
+                case COMMUNICATION_ERROR:
+                case ONLINE:
+                    if (!loginSuccess) {
+                        login(detectionAction, DETECT_TIMEOUT_MS);
+                    }
+
+                    if (!getLastDetection(false)) {
+                        // Try login again in case the session has timed out
+                        login(detectionAction, DETECT_TIMEOUT_MS);
+                        getLastDetection(true);
+                    }
+                    break;
+                default:
+                    break;
+            }
+
+            if (loginSuccess && detectSuccess) {
+                status = DeviceStatus.ONLINE;
+                if (!online) {
+                    online = true;
+                    listener.sensorStatus(status);
+
+                    // Ignore old detections
+                    prevDetection = lastDetection;
+                }
+
+                if (lastDetection != prevDetection) {
+                    listener.motionDetected();
+                }
+            } else {
+                if (online || updateStatus) {
+                    online = false;
+                    listener.sensorStatus(status);
+                }
+            }
+        }
+    };
+
+    public DLinkMotionSensorCommunication(final DLinkMotionSensorConfig config,
+            final DLinkMotionSensorListener listener, final ScheduledExecutorService scheduler) {
+        super(config.ipAddress, config.pin);
+        this.listener = listener;
+
+        if (getHNAPStatus() == HNAPStatus.INTERNAL_ERROR) {
+            status = DeviceStatus.INTERNAL_ERROR;
+        }
+
+        try {
+            final MessageFactory messageFactory = MessageFactory.newInstance();
+            detectionAction = messageFactory.createMessage();
+
+            buildDetectionAction();
+
+        } catch (final SOAPException e) {
+            logger.debug("DLinkMotionSensorCommunication - Internal error", e);
+            status = DeviceStatus.INTERNAL_ERROR;
+        }
+
+        detectFuture = scheduler.scheduleWithFixedDelay(detect, 0, DETECT_POLL_S, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Stop communicating with the device
+     */
+    @Override
+    public void dispose() {
+        detectFuture.cancel(true);
```

Comment:
detectFuture is initialised in the constructor so I wouldn't expect it to ever be null.